doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+
+## Context
+
+We want to make it easier for folks to get data into Materialize. Currently the only way is via one
+of our [supported sources](https://materialize.com/docs/sql/create-source/), which depending on
+the environment, can be pretty hard to setup. A common method of publishing data from an
+application is via [webhooks](https://www.redhat.com/en/topics/automation/what-is-a-webhook). When
+some event occurrs, e.g. a new analytics event from Segment, or a new Pull Request on a GitHub
+repository, an HTTP endpoint you define will get called so you can receive the event in near
+real-time.
+
+Supporting webhooks in Materialize allows users to easily get their data into MZ, and the low
+latency of the events works very well with the premise of being an "operational" data warehouse.
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to. In Materialize this data is represented as a source
+with two columns, `headers` and `body`.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM HTTP REQUESTS` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `HTTP REQUESTS` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Validation that the request comes from a trusted source.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `HTTP REQUESTS` source, and add
+a new endpoint to `environmentd`'s HTTP server `/api/webhook/:id`. When a user creates this new
+kind of source, we'll generate an id, persist the new `Source` in the catalog, use the
+`storage-client` to create a new collection, and finally return a well formatted URL to the user.
+
+When we receive a event we'll use the created `SessionClient` to send the raw `headers` and `body`
+to the coordinator, pack the data into a row, and append it to the necessary collection using a
+new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `HTTP REQUESTS`, for which you can
+specify a `FORMAT`, which denotes how we decode the body of a request, the only supported format
+for the time being is `BYTES`.
```

Comment:
I'd like to counter-propose for the initial scope to instead be either `FORMAT JSON` or both JSON and BYTES

all of these webhook sources will be append-only and if the customer has any issues with dataflow restart times, we won't have any sort of story for them (b/c as called out below, pushdown wouldn't work on BYTES). even if we then built support for `FORMAT JSON` webhooks, AFAIK we wouldn't have a story for migrating the data (unlike kafka+postgres, dropping and recreating this source would clear it out). given that it's mentioned elsewhere that all our initial uses of this are json and that writing things down in BYTES has material downsizes, this feels to me like a premature generalization

Reply:
+1 to everything Dan said

Reply:
Supporting `FORMAT JSON` in the initial implementation makes sense to me. I'm actually not sure why we don't support it already for Kafka topics!

Started a conversation about this here: https://materializeinc.slack.com/archives/C01CFKM1QRF/p1687298014509359

Reply:
Makes sense to me! I updated the doc to include `JSON`, `BYTES`, and a `TEXT` format. I don't think `TEXT` is strictly necessary, but adding it is virtually 0 work once if we're already supporting `JSON` and `BYTES`